zkevm-circuits/src/super_circuit.rs

```diff
@@ -0,0 +1,411 @@
+            let tx_rlp = req.rlp(chain_id);
+            let sighash = keccak256(tx_rlp.as_ref()).into();
+            let sig = wallet.sign_hash(sighash, true);
+            tx.v = U64::from(sig.v);
+            tx.r = sig.r;
+            tx.s = sig.s;
+        }
+    }
+}
+
+#[cfg(test)]
+mod super_circuit_tests {
+    use super::test::*;
+    use super::*;
+    use crate::{evm_circuit::witness::block_convert, tx_circuit::sign_verify::POW_RAND_SIZE};
+    use bus_mapping::mock::BlockData;
+    use eth_types::{
+        address, bytecode,
+        geth_types::{self, GethData},
+        Word,
+    };
+    use ethers_signers::{LocalWallet, Signer};
+    use group::{Curve, Group};
+    use halo2_proofs::arithmetic::{CurveAffine, Field as Halo2Field};
+    use halo2_proofs::dev::{MockProver, VerifyFailure};
+    use halo2_proofs::pairing::bn256::Fr;
+    use mock::{TestContext, MOCK_CHAIN_ID};
+    use rand::SeedableRng;
+    use rand_chacha::ChaCha20Rng;
+    use secp256k1::Secp256k1Affine;
+    use std::collections::HashMap;
+    use strum::IntoEnumIterator;
+
+    struct Inputs<F: Field> {
+        block: Block<F>,
+        txs: Vec<geth_types::Transaction>,
+        aux_generator: Secp256k1Affine,
+    }
+
+    fn run_test_circuit<F: Field, const MAX_TXS: usize, const MAX_CALLDATA: usize>(
+        inputs: Inputs<F>,
+        fixed_table_tags: Vec<FixedTableTag>,
+    ) -> Result<(), Vec<VerifyFailure>> {
+        let Inputs {
+            block,
+            txs,
+            aux_generator,
+        } = inputs;
+
+        let log2_ceil = |n| u32::BITS - (n as u32).leading_zeros() - (n & (n - 1) == 0) as u32;
+
+        let num_rows_required_for_steps =
+            SuperCircuit::<F, MAX_TXS, MAX_CALLDATA>::get_num_rows_required(&block);
+
+        let k = log2_ceil(
+            64 + fixed_table_tags
+                .iter()
+                .map(|tag| tag.build::<F>().count())
+                .sum::<usize>(),
+        );
+        let bytecodes_len = block
+            .bytecodes
+            .iter()
+            .map(|(_, bytecode)| bytecode.bytes.len())
+            .sum::<usize>();
+        let k = k.max(log2_ceil(64 + bytecodes_len));
+        let k = k.max(log2_ceil(64 + num_rows_required_for_steps));
+        let k = k + 1;
+        log::debug!("evm circuit uses k = {}", k);
+
+        let mut instance: Vec<Vec<F>> = (1..POW_RAND_SIZE + 1)
+            .map(|exp| vec![block.randomness.pow(&[exp as u64, 0, 0, 0]); (1 << k) - 64])
+            .collect();
+        // SignVerifyChip -> ECDSAChip -> MainGate instance column
+        instance.push(vec![]);
+
+        let chain_id = block.context.chain_id;
+        let tx_circuit = TxCircuit::new(aux_generator, block.randomness, chain_id.as_u64(), txs);
+        let circuit = SuperCircuit::<F, MAX_TXS, MAX_CALLDATA> {
+            block,
+            fixed_table_tags,
+            tx_circuit,
+            bytecode_size: bytecodes_len + 64,
+        };
```

Comment:
Right, originally I had `1 << k` here, so that the bytecode region occupies the entire circuit; but the other circuits require a big k, which ends up being `k = 19`.  And with `2^19` enabled rows in the bytecode circuit, the mock prover takes 6h for a test run!
So that's why I decided to use enough rows to hold the bytecode in the test, plus some extra rows so that we also check the constraints for padding rows in the circuit.

Reply:
I see! It makes sense now.